### ISSUE TITLE
Default shuffle behavior of DistributedSampler

### ISSUE THREAD

**Issue by viperit** [pytorch/pytorch#31232]
## 🚀 Feature
The **DistributedSampler** should shuffle data without explicitly set `set_epoch`.

## Motivation
I have found that the parameter shuffle of **DistributedSampler** is set True in default but the behavior and intuition are inconsistent. The GPU always load the same data between every epoch. It still needs manual set sampler.set_epoch to change the data loaded by GPU
## Pitch
Truely shuffle the data when the shuffle=True in **DistributedSampler** 
## Alternatives
Just show this potential feature in doc.
## Additional context

The test code, 

`CUDA_VISIBLE_DEVICES=1,3 python -m torch.distributed.launch --nproc_per_node=2 test.py`

test.py:
```import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torch.utils.data.distributed import DistributedSampler


torch.distributed.init_process_group(backend="nccl")

input_size = 5
output_size = 2
batch_size = 2
data_size = 16

local_rank = torch.distributed.get_rank()
torch.cuda.set_device(local_rank)
device = torch.device("cuda", local_rank)

class RandomDataset(Dataset):
    def __init__(self, size, length, local_rank):
        self.len = length
        self.data = torch.stack([torch.ones(5), torch.ones(5)*2,
                                 torch.ones(5)*3,torch.ones(5)*4,
                                 torch.ones(5)*5,torch.ones(5)*6,
                                 torch.ones(5)*7,torch.ones(5)*8,
                                 torch.ones(5)*9, torch.ones(5)*10,
                                 torch.ones(5)*11,torch.ones(5)*12,
                                 torch.ones(5)*13,torch.ones(5)*14,
                                 torch.ones(5)*15,torch.ones(5)*16]).to('cuda')

        self.local_rank = local_rank
    def __getitem__(self, index):

        return self.data[index]

    def __len__(self):
        return self.len
    
dataset = RandomDataset(input_size, data_size, local_rank)
sampler = DistributedSampler(dataset)
rand_loader = DataLoader(dataset=dataset,
                         batch_size=batch_size,
                         sampler=sampler)

e = 0
while e < 2:
    t = 0
    sampler.set_epoch(e)
    for data in rand_loader:
        print(data)
    e+=1
```

**Comment by ydaiming** [comment]
Hi there, if you haven't, I'm going to publish PR for this small twist.

**Comment by zhaojuanmao** [comment]
@ydaiming not yet, feel free to send a PR, we will review it. Thanks for contributing
